# Chapter 1: Python Basics

## 1.1 Using Python

In [1]:
# check if you have python installed
print("Hello, Universe!")

Hello, Universe!


## 1.2 Understanding Expression and Assignments

In [6]:
# calculate Earth velocity around the Sun
# 2*Pi*radius in km/period which is a year in seconds
print(2 * 3.14159 * (1.496* 10**8) / (3.156*10**7))
# or 
print(2 * 3.14159 * 1.496e8/ 3.156e7)

29.783388086185045
29.783388086185045


In [11]:
# make a simple calculator
radius = 1.496e8 # orbital radius in km
period = 3.156e7 # orbital period in s
pi = 3.14159     # value of pi

# calculate the orbital velocity
velocity = 2 * pi * radius / period

# display output
print(velocity)
# or
print("orbital velocity = ", velocity, " km/s")

29.783388086185045
orbital velocity =  29.783388086185045  km/s


In [12]:
# discard unimportant digits that might cause errors in computations
print("orbital velocity = {:5.2f} km/s".format(velocity))

orbital velocity = 29.78 km/s


In [13]:
# calculate the radius of an imaginary planet that is 10 times bigger than Earth
radius = 10*radius
print("the radius of the imaginary planet = {:.3e} km".format(radius))


the radius of the imaginary planet = 1.496e+09 km


### Kepler's third law of planetary motion, for a planet that orbits in a circular motion around our sun
$${P^2} ={\frac{4\pi^2}{GM}}\times{r^3}$$
where $$M = 1.989\times{10^30}kg$$ is the mass of the ☀
and $$G = 6.674\times{10^-10} N {kg^-2}{m^2}$$ is the gravitational constant
we can rewrite it as: $$P = 2\pi \times{GM^{\frac{-1}{2}}}{r^{\frac{3}{2}}}$$



In [14]:
# calculator for period in seconds given the radius in km based on Kepler's third law 
g = 6.674e-11
m = 1.989e30
period = 2 * pi * (g * m ) ** (-1/2) * (1e3 * radius) ** (3/2) # we converted the radius from km to m

# display period on years
print("new orbital period = {:.1f} years".format(period/3.156e7)) # we converted seconds into years

velocity = 2 * pi * radius / period
# dispaly velocity
print("new orbital velocity = {:.2f} km/s".format(velocity))
# remember that this velocity belongs to that imaginary planet that is 10 times bigger than Earth!


new orbital period = 31.6 years
new orbital velocity = 9.42 km/s


### so 1 year in this planet is 31.6 of Earth years! as its slower than earth when it orbits around the ☀

## 1.3 Control Structures

In [1]:
# calculating the sum of first 100 natural numbers using for loop
sum = 0 # give initial value for the sum
iterations = 100 # number of iterations or how many times we repeat the loop

for k in range (1, iterations + 1): # for every item in range from 1 until last iteration, add sum
    sum = sum + k

print("sum of the first 100 natural numbers is ", sum)

sum of the first 100 natural numbers is  5050


#### Another solution found by Gauss:
$$ sum = \frac {n(n-1)}{2}

In [6]:
n = 100 + 1
sum = (n*(n - 1)) / 2

print("sum of the first 100 natural numbers is ", sum)


sum of the first 100 natural numbers is  5050.0


### Fibonacci sequence
$$ F_n = F_{n-1} + F_{n-2}$$
where $$ F_0 = 0 $$ and $$ F_1 = 1 $$


In [9]:
# fibonacci calculator
n_max = 10 # compute the sequence until we reach this item

# initialize the constants
f_prev = 0
f = 1
for n in range (1, n_max+1): # compute until we reach the last item
    print("{:d}. fibonacci number is {:d}".format(n, f))
    f_next = f + f_prev # sume the current item with the item before it
    f_prev = f # make the current value as the old value
    f = f_next # update the next to be the current value

1. fibonacci number is 1
2. fibonacci number is 1
3. fibonacci number is 2
4. fibonacci number is 3
5. fibonacci number is 5
6. fibonacci number is 8
7. fibonacci number is 13
8. fibonacci number is 21
9. fibonacci number is 34
10. fibonacci number is 55


### rewrite it adding a base condition

In [11]:
f_prev = 0
f, n = 1, 1

# compute the sequence until 1000
while f < 1000:
    print("{:d}. fibonacci number is {:d}".format(n, f))
    f_next = f + f_prev # sume the current item with the item before it
    f_prev = f # make the current value as the old value
    f = f_next # update the next to be the current value
    n+=1 # increment counter of the loop
    

1. fibonacci number is 1
2. fibonacci number is 1
3. fibonacci number is 2
4. fibonacci number is 3
5. fibonacci number is 5
6. fibonacci number is 8
7. fibonacci number is 13
8. fibonacci number is 21
9. fibonacci number is 34
10. fibonacci number is 55
11. fibonacci number is 89
12. fibonacci number is 144
13. fibonacci number is 233
14. fibonacci number is 377
15. fibonacci number is 610
16. fibonacci number is 987


#### Test if the calculated value is even or odd

In [13]:
f_prev = 0
f = 1
n_odd = 0
n_even = 0
# compute the sequence until 1000
while f < 1000:
    f_next = f + f_prev # sume the current item with the item before it
    f_prev = f # make the current value as the old value
    f = f_next # update the next to be the current value

    if f%2 == 0:
        n_even +=1
    else:
        n_odd +=1
print("Found {:d} even numbers and {:d} odd numbers".format(n_even, n_odd))


Found 5 even numbers and 11 odd numbers


## 1.4 Working with Modules and Objects

In [14]:
# explore some predefined physical constants
import scipy.constants
dir(scipy.constants)

['Avogadro',
 'Boltzmann',
 'Btu',
 'Btu_IT',
 'Btu_th',
 'ConstantWarning',
 'G',
 'Julian_year',
 'N_A',
 'Planck',
 'R',
 'Rydberg',
 'Stefan_Boltzmann',
 'Wien',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_obsolete_constants',
 'acre',
 'alpha',
 'angstrom',
 'arcmin',
 'arcminute',
 'arcsec',
 'arcsecond',
 'astronomical_unit',
 'atm',
 'atmosphere',
 'atomic_mass',
 'atto',
 'au',
 'bar',
 'barrel',
 'bbl',
 'blob',
 'c',
 'calorie',
 'calorie_IT',
 'calorie_th',
 'carat',
 'centi',
 'codata',
 'constants',
 'convert_temperature',
 'day',
 'deci',
 'degree',
 'degree_Fahrenheit',
 'deka',
 'dyn',
 'dyne',
 'e',
 'eV',
 'electron_mass',
 'electron_volt',
 'elementary_charge',
 'epsilon_0',
 'erg',
 'exa',
 'exbi',
 'femto',
 'fermi',
 'find',
 'fine_structure',
 'fluid_ounce',
 'fluid_ounce_US',
 'fluid_ounce_imp',
 'foot',
 'g',
 'gallon',
 'gallon_US',
 'gallon_imp',
 'gas_constant'

In [15]:
print(scipy.constants.gravitational_constant)

6.6743e-11


In [16]:
import scipy.constants as const
print(const.G)
# or
from scipy.constants import G
print(G)

6.6743e-11
6.6743e-11


#### Modify the orbital velocity calculator using the provided constants

In [18]:
from math import pi, sqrt
from astropy.constants import M_sun
from scipy.constants import au, G, year

print(" 1 astronomical unit is ", au, " m")
print(" 1 year is ", year, " s")

radius = 10 * au
print(" radial distance is {:.1f} au".format(radius/au))

# Kepler's third law of planetary motion
period = 2 * pi * sqrt(radius ** 3 / (G * M_sun.value))
print(" orbital period is {:.4f} years".format(period/year))

velocity = 2 * pi * radius / period
print(" orbital velocity is {:.2f} km/s".format(1e-3*velocity))


 1 astronomical unit is  149597870700.0  m
 1 year is  31536000.0  s
 radial distance is 10.0 au
 orbital period is 31.6450 years
 orbital velocity is 9.42 km/s


In [19]:
print(M_sun)

  Name   = Solar mass
  Value  = 1.988409870698051e+30
  Uncertainty  = 4.468805426856864e+25
  Unit  = kg
  Reference = IAU 2015 Resolution B 3 + CODATA 2018


In [20]:
from astropy.constants import M_earth
print(M_earth)

  Name   = Earth mass
  Value  = 5.972167867791379e+24
  Uncertainty  = 1.3422009501651213e+20
  Unit  = kg
  Reference = IAU 2015 Resolution B 3 + CODATA 2018
